src/Typechecker.ml

```diff
@@ -49,6 +49,15 @@ let rec check_open delta gamma Location.{ data = expr; loc } typ =
       | _ ->
           fail_in loc
           @@ `TypeMismatchError "snd is applied to a non-product type" )
+  | Nat _ ->
+      with_error_location loc @@ check_equal typ Type.Nat "Expected nat type"
+  | BinOp { op = _; e1; e2 } ->
+      let%bind ty1 = infer_open delta gamma e1 in
+      let%bind ty2 = infer_open delta gamma e2 in
+      let%bind () =
+        with_error_location loc @@ check_equal ty1 Type.Nat "Expected nat type"
+      in
+      with_error_location loc @@ check_equal ty2 Type.Nat "Expected nat type"
```

Comment:
In this case, we know exactly what the type of each subexpression must be, so we need not infer those types but *check* them. So it should something like this:
```suggestion
      let%map () = check_open delta gamma e1 Type.Nat
      and () = check_open delta gamma e2 Type.Nat in
      ()
```